Mathlib/RingTheory/SurjectiveOnStalks.lean

```diff
@@ -0,0 +1,181 @@
+        IsLocalization.mk'_eq_iff_eq, IsLocalization.eq_iff_exists P.primeCompl]
+      refine ⟨⟨yc, hyc⟩ * ⟨yt, hyt⟩, ?_⟩
+      simp only [Submonoid.coe_mul]
+      convert congr($(ey.symm) * $(et)) using 1 <;> ring
+
+lemma surjectiveOnStalks_iff_forall_ideal :
+    f.SurjectiveOnStalks ↔
+      ∀ I : Ideal S, I ≠ ⊤ → ∀ s : S, ∃ x r : R, ∃ c ∉ I, f r ∉ I ∧ c * f r * s = c * f x := by
+  simp_rw [SurjectiveOnStalks, surjective_localRingHom_iff]
+  refine ⟨fun H I hI s ↦ ?_, fun H I hI ↦ H I hI.ne_top⟩
+  obtain ⟨M, hM, hIM⟩ := I.exists_le_maximal hI
+  obtain ⟨x, r, c, hc, hr, e⟩ := H M hM.isPrime s
+  exact ⟨x, r, c, fun h ↦ hc (hIM h), fun h ↦ hr (hIM h), e⟩
+
+lemma surjectiveOnStalks_iff_forall_maximal :
+    f.SurjectiveOnStalks ↔ ∀ (I : Ideal S) (_ : I.IsMaximal),
+      Function.Surjective (Localization.localRingHom _ I f rfl) := by
+  refine ⟨fun H I hI ↦ H I hI.isPrime, fun H I hI ↦ ?_⟩
+  simp_rw [surjective_localRingHom_iff] at H ⊢
+  intro s
+  obtain ⟨M, hM, hIM⟩ := I.exists_le_maximal hI.ne_top
+  obtain ⟨x, r, c, hc, hr, e⟩ := H M hM s
+  exact ⟨x, r, c, fun h ↦ hc (hIM h), fun h ↦ hr (hIM h), e⟩
+
+lemma surjectiveOnStalks_iff_forall_maximal' :
+    f.SurjectiveOnStalks ↔ ∀ I : Ideal S, I.IsMaximal →
+      ∀ s : S, ∃ x r : R, ∃ c ∉ I, f r ∉ I ∧ c * f r * s = c * f x := by
+  simp only [surjectiveOnStalks_iff_forall_maximal, surjective_localRingHom_iff]
+
+lemma surjectiveOnStalks_of_exists_div (h : ∀ x : S, ∃ r s : R, IsUnit (f s) ∧ f s * x = f r) :
+    SurjectiveOnStalks f :=
+  surjectiveOnStalks_iff_forall_ideal.mpr fun I hI x ↦
+    let ⟨r, s, hr, hr'⟩ := h x
+    ⟨r, s, 1, by simpa [← Ideal.eq_top_iff_one], fun h ↦ hI (I.eq_top_of_isUnit_mem h hr), by simpa⟩
+
+lemma surjectiveOnStalks_of_surjective (h : Function.Surjective f) :
+    SurjectiveOnStalks f :=
+  surjectiveOnStalks_iff_forall_ideal.mpr fun _ _ s ↦
+    let ⟨r, hr⟩ := h s
+    ⟨r, 1, 1, by simpa [← Ideal.eq_top_iff_one], by simpa [← Ideal.eq_top_iff_one], by simp [hr]⟩
+
+variable (S) in
+lemma surjectiveOnStalks_of_isLocalization [IsLocalization M S] :
+    SurjectiveOnStalks (algebraMap R S) := by
+  refine surjectiveOnStalks_of_exists_div fun s ↦ ?_
+  obtain ⟨x, s, rfl⟩ := IsLocalization.mk'_surjective M s
+  exact ⟨x, s, IsLocalization.map_units S s, IsLocalization.mk'_spec' S x s⟩
+
+lemma SurjectiveOnStalks.comp (hg : SurjectiveOnStalks g) (hf : SurjectiveOnStalks f) :
+    SurjectiveOnStalks (g.comp f) := by
+  intros I hI
+  have := (hg I hI).comp (hf _ (hI.comap g))
+  rwa [← RingHom.coe_comp, ← Localization.localRingHom_comp] at this
+
+lemma SurjectiveOnStalks.of_comp (hg : SurjectiveOnStalks (g.comp f)) :
+    SurjectiveOnStalks g := by
+  intros I hI
+  have := hg I hI
+  rw [Localization.localRingHom_comp (I.comap (g.comp f)) (I.comap g) _ _ rfl _ rfl,
+    RingHom.coe_comp] at this
+  exact this.of_comp
+
+open TensorProduct
+
+/--
+If `R → T` is surjective on stalks, and `J` is some prime of `T`,
```

Comment:
The docstring doesn't make sense.
`(1 ⊗ r • t) * x = a ⊗ t` for some `a : S`, `r : R` and `t: T` such that `r • t ∉ J`